File: Graph/csharp/BestBridge.cs
using System.Collections.Generic;

namespace GraphSolutions;

public static class BestBridge
{
    public static int Solve(char[][] grid)
    {
        var firstIsland = new HashSet<(int, int)>();
        var found = false;

        for (var r = 0; r < grid.Length && !found; r++)
        {
            for (var c = 0; c < grid[0].Length && !found; c++)
            {
                if (grid[r][c] == 'L')
                {
                    CollectIsland(grid, r, c, firstIsland);
                    if (firstIsland.Count > 0)
                    {
                        found = true;
                    }
                }
            }
        }

        if (firstIsland.Count == 0)
        {
            return 0;
        }

        var visited = new HashSet<(int, int)>(firstIsland);
        var queue = new Queue<(int Row, int Col, int Distance)>();

        foreach (var (row, col) in firstIsland)
        {
            queue.Enqueue((row, col, 0));
        }

        while (queue.Count > 0)
        {
            var (row, col, distance) = queue.Dequeue();
            var current = (row, col);

            if (grid[row][col] == 'L' && !firstIsland.Contains(current))
            {
                return distance - 1;
            }

            foreach (var (dr, dc) in GraphUtils.Directions)
            {
                var nr = row + dr;
                var nc = col + dc;
                var neighbor = (nr, nc);
                if (!GraphUtils.InBounds(grid, nr, nc) || !visited.Add(neighbor))
                {
                    continue;
                }

                queue.Enqueue((nr, nc, distance + 1));
            }
        }

        return -1;
    }

    private static void CollectIsland(char[][] grid, int row, int col, HashSet<(int, int)> island)
    {
        if (!GraphUtils.InBounds(grid, row, col) || grid[row][col] == 'W')
        {
            return;
        }

        var pos = (row, col);
        if (!island.Add(pos))
        {
            return;
        }

        foreach (var (dr, dc) in GraphUtils.Directions)
        {
            CollectIsland(grid, row + dr, col + dc, island);
        }
    }
}

File: Graph/java/BestBridge.java
import java.util.ArrayDeque;
import java.util.HashSet;
import java.util.Queue;
import java.util.Set;

public class BestBridge {
  public static int bestBridge(char[][] grid) {
    Set<String> firstIsland = new HashSet<>();

    outer:
    for (int r = 0; r < grid.length; r++) {
      for (int c = 0; c < grid[0].length; c++) {
        if (grid[r][c] == 'L') {
          collect(grid, r, c, firstIsland);
          if (!firstIsland.isEmpty()) {
            break outer;
          }
        }
      }
    }

    if (firstIsland.isEmpty()) {
      return 0;
    }

    Queue<int[]> queue = new ArrayDeque<>();
    Set<String> visited = new HashSet<>(firstIsland);

    for (String key : firstIsland) {
      String[] parts = key.split(",");
      int row = Integer.parseInt(parts[0]);
      int col = Integer.parseInt(parts[1]);
      queue.add(new int[] {row, col, 0});
    }

    while (!queue.isEmpty()) {
      int[] state = queue.poll();
      int row = state[0];
      int col = state[1];
      int distance = state[2];
      String key = row + "," + col;

      if (grid[row][col] == 'L' && !firstIsland.contains(key)) {
        return distance - 1;
      }

      for (int[] delta : GraphUtils.DIRECTIONS) {
        int nr = row + delta[0];
        int nc = col + delta[1];
        String neighborKey = nr + "," + nc;
        if (!GraphUtils.inBounds(grid, nr, nc) || visited.contains(neighborKey)) {
          continue;
        }
        visited.add(neighborKey);
        queue.add(new int[] {nr, nc, distance + 1});
      }
    }

    return -1;
  }

  private static void collect(char[][] grid, int row, int col, Set<String> island) {
    if (!GraphUtils.inBounds(grid, row, col) || grid[row][col] == 'W') {
      return;
    }

    String key = row + "," + col;
    if (!island.add(key)) {
      return;
    }

    for (int[] delta : GraphUtils.DIRECTIONS) {
      collect(grid, row + delta[0], col + delta[1], island);
    }
  }
}
